File: test_euler.py
import math

# Matrix transposed because Three.js uses column-major arrays, but the numbers I provided are row-by-row visually
# Wait, no, Three.js set() takes row-major arguments!
# so m.elements in Three.js is column-major.
# Let's just do it manually.
# [ m00 m01 m02 m03 ]
# [ m10 m11 m12 m13 ]
# [ m20 m21 m22 m23 ]
# [ m30 m31 m32 m33 ]

m00 = 0.9894; m01 = -0.1385; m02 = -0.0429
m10 = -0.000; m11 = -0.2962; m12 =  0.9551
m20 = -0.1450; m21 = -0.9450; m22 = -0.2931

# Assuming YXZ order like Three.js default
# Y = asin( clamp( m02, -1, 1 ) )? No, Three.js Euler setFromRotationMatrix
# Pitch is X, Yaw is Y, Roll is Z

# Let's just output the matrix and check rotation around X
pitch = math.atan2(-m12, m22)
print("Pitch (atan2(-m12, m22)):", pitch * 180 / math.pi)

pitch2 = math.asin(m12)
print("Pitch (asin(m12)):", pitch2 * 180 / math.pi)

yaw = math.atan2(m02, m22)
print("Yaw (atan2(m02, m22)):", yaw * 180 / math.pi)

File: test_euler.js
const THREE = require('three');

const m = new THREE.Matrix4().set(
  0.9894, -0.1385, -0.0429, 0.0,
  -0.000, -0.2962,  0.9551, 0.0,
 -0.1450, -0.9450, -0.2931, 0.0,
  0.0,     0.0,     0.0,    1.0
);

const euler = new THREE.Euler().setFromRotationMatrix(m, 'YXZ');
console.log("Original YXZ:", euler.x * 180/Math.PI, euler.y * 180/Math.PI, euler.z * 180/Math.PI);

euler.x = 0; // Remove pitch? Wait, in YXZ, X is pitch. But what if we want to remove the pitch RELATIVE to the world?
const newM = new THREE.Matrix4().makeRotationFromEuler(euler);

console.log("New Matrix Y-axis:", newM.elements[4], newM.elements[5], newM.elements[6]);
console.log("New Matrix Z-axis:", newM.elements[8], newM.elements[9], newM.elements[10]);

// Try another order
const eulerZYX = new THREE.Euler().setFromRotationMatrix(m, 'ZYX');
console.log("Original ZYX:", eulerZYX.x * 180/Math.PI, eulerZYX.y * 180/Math.PI, eulerZYX.z * 180/Math.PI);
